lib.py

```diff
@@ -1,167 +1,284 @@
-
-        cur.execute(
-            "INSERT INTO runs VALUES (?, ?, ?, ?, ?, ?, ?)",
-            (
-                str(msg.author.id),
-                code,
-                day,
-                part,
-                result["median"],
-                result["answer"],
-                result["answer"],
-            ),
-        )
-        results.append(result)
+def process_run_result(in_file, answers_map, result_lst):
+    """Given JSON blobs extracted from a container's stdout, get the core stats out."""
+    # `result` should be a dataclass, not a dict. Another time.
+    result = {}
+    for blob in result_lst:
+        reason = blob.get("reason", None)
+        if reason == None:
+            continue
+        elif reason == "ferris-answer":
+            answer = blob["answer"]
+            result["answer"] = answer
+            if answers_map.get(in_file, None) == answer:
+                result["verified"] = True
+            else:
+                result["verified"] = False
+        elif reason == "benchmark-complete":
+            result["typical"] = blob["typical"]["estimate"]
+            result["average"] = blob["mean"]["estimate"]
+            result["median"] = blob["median"]["estimate"]
+            result["high_bound"] = blob["typical"]["upper_bound"]
+            result["low_bound"] = blob["typical"]["lower_bound"]
+    logger.info("Computed run result: %s", result)
+    return result
 
-    # TODO: Make this conditional on data actually being inserted into the DB
-    db.commit()
-    logger.info("Inserted %s results into the DB", len(results))
+def save_results(cur: Cursor, author_id: int, day: int, part:int, code: bytes, results):
+    """
+    Save the benchmark run results to the DB.
+    """
 
-    median = avg([r["median"] for r in results])
-    average = avg([r["average"] for r in results])
+    str_code = code.decode()
+    db_results = [(str(author_id), str_code, day, part, r["median"], int(r["answer"]) if r["answer"].isdigit() else None, r["answer"]) for r in results]
 
-    if verified:
-        await msg.reply(
-            embed=discord.Embed(
-                title="Benchmark complete",
-                description=f"Median: **{ns(median)}**\nAverage: **{ns(average)}**",
-            )
-        )
-    else:
-        await msg.reply(
-            embed=discord.Embed(
-                title="Benchmark complete (Unverified)",
-                description=f"Median: **{ns(median)}**\nAverage: **{ns(average)}**",
-            )
-        )
+    query = "INSERT INTO runs(user, code, day, part, time, answer, answer2) VALUES (?, ?, ?, ?, ?, ?, ?)"
+
+    cur.executemany(query, db_results)
 
+def ns(v):
+    """Format number of nanoseconds for display."""
+    if v > 1e9:
+        return f"{v / 1e9:.2f}s"
+    if v > 1e6:
+        return f"{v / 1e6:.2f}ms"
+    if v > 1e3:
+        return f"{v / 1e3:.2f}µs"
+    return f"{v:.2f}ns"
```

Comment:
a note on this, i think we should move to picosecond accuracy for our benches, since some days enter the ns range and its nice to have the extra accuracy, it only takes 40 bits if we cap the maximum time to 1 second (well within the 63 bits needed for sqlite)

I'll discuss this more in a issue/PR to refactor the database design to be more efficient overall though, my main comment here is since we use nanosecond accuracy we shouldn't bother to render the `{v:.2f}ns` with 2 digits after the decimal place, since they are always zero

Reply:
OK, yeah, fair point. I'm somewhat skeptical about our ability to measure with picosecond precision, but maybe we can. Either way, we can definitely change the format string.

Reply:
I should clarify: We can nail down specifics of of stored precision & database format in the off-season.